File: backend/app/core/booking_window.py
from datetime import date, datetime, time, timedelta
from typing import Final
from zoneinfo import ZoneInfo

_ROME_TIMEZONE: Final[ZoneInfo] = ZoneInfo("Europe/Rome")

# Monday=0 .. Sunday=6, per date.weekday().
_UNLOCK_WEEKDAY: Final[int] = 4
_UNLOCK_HOUR: Final[int] = 20
_CURRENT_WEEK_LAST_OFFSET: Final[int] = 6
_NEXT_WEEK_LAST_OFFSET: Final[int] = 13

_OUT_OF_WINDOW_ERROR: Final[str] = (
    "La data selezionata non rientra nella finestra di prenotazione "
    "attualmente sbloccata."
)


# Rolling weekly unlock: every Friday at 20:00 the whole following week opens
# up, on top of whatever is left of the already unlocked current one.
def compute_available_window(now: datetime) -> tuple[date, date]:
    today = now.date()
    current_week_monday = today - timedelta(days=today.weekday())
    unlock_moment = datetime.combine(
        current_week_monday + timedelta(days=_UNLOCK_WEEKDAY),
        time(_UNLOCK_HOUR),
        tzinfo=now.tzinfo,
    )
    last_unlocked_offset = (
        _NEXT_WEEK_LAST_OFFSET
        if now >= unlock_moment
        else _CURRENT_WEEK_LAST_OFFSET
    )

    return today, current_week_monday + timedelta(days=last_unlocked_offset)


def today_in_rome() -> date:
    return datetime.now(_ROME_TIMEZONE).date()


# The ValueError is mapped to HTTP 400 by the app-wide handler.
def assert_within_booking_window(target_date: date) -> None:
    today, max_unlocked = compute_available_window(datetime.now(_ROME_TIMEZONE))

    if target_date < today or target_date > max_unlocked:
        raise ValueError(_OUT_OF_WINDOW_ERROR)

File: frontend/lib/features/lessons/utils/booking_window.dart
export '../../../core/utils/week_range.dart' show isSameDate;

// Weekly unlock window for availabilities and presences: every Friday at
// 20:00 the entire following week (Monday-Sunday) opens up, on top of
// whatever remains of the already unlocked current week. Mirrors
// backend/app/core/booking_window.py — keep both in sync if the rule changes.
//
// Every day-stepping computation here goes through the DateTime constructor
// (year, month, day + n) rather than Duration-based add()/subtract(): Duration
// arithmetic on local DateTimes can skew across Italy's DST transitions,
// while constructor field arithmetic is plain calendar math with no such risk.

const List<String> _weekdayNames = [
  'Lunedì',
  'Martedì',
  'Mercoledì',
  'Giovedì',
  'Venerdì',
  'Sabato',
  'Domenica',
];

const List<String> _monthNames = [
  'gennaio',
  'febbraio',
  'marzo',
  'aprile',
  'maggio',
  'giugno',
  'luglio',
  'agosto',
  'settembre',
  'ottobre',
  'novembre',
  'dicembre',
];

const int _unlockHour = 20;

DateTime _dateOnly(DateTime value)
{
  return DateTime(value.year, value.month, value.day);
}

DateTime _addDays(DateTime date, int days)
{
  return DateTime(date.year, date.month, date.day + days);
}

DateTime _mondayOf(DateTime day)
{
  return _addDays(day, -(day.weekday - DateTime.monday));
}

// Whether the week after [now]'s has opened: from Friday 20:00 to Sunday night.
bool isNextWeekUnlocked(DateTime now)
{
  final currentWeekMonday = _mondayOf(_dateOnly(now));
  final fridayThisWeek20 = DateTime(
    currentWeekMonday.year,
    currentWeekMonday.month,
    currentWeekMonday.day + (DateTime.friday - DateTime.monday),
    _unlockHour,
  );

  return !now.isBefore(fridayThisWeek20);
}

// The last unlocked day: this week's Sunday, or next week's once it opens.
DateTime lastUnlockedDay(DateTime now)
{
  final currentWeekMonday = _mondayOf(_dateOnly(now));

  return _addDays(currentWeekMonday, isNextWeekUnlocked(now) ? 13 : 6);
}

// Today through the furthest unlocked day, inclusive; never empty (3 to 10 days).
List<DateTime> computeAvailableDays(DateTime now)
{
  final today = _dateOnly(now);
  final maxUnlocked = lastUnlockedDay(now);

  final days = <DateTime>[];
  var cursor = today;

  while (!cursor.isAfter(maxUnlocked))
  {
    days.add(cursor);
    cursor = _addDays(cursor, 1);
  }

  return days;
}

// E.g. "Sabato 25 luglio". No locale data is initialized in this app, so
// DateFormat with a locale would throw at runtime; manual lookup instead.
String formatAvailableDayLabel(DateTime date)
{
  final weekday = _weekdayNames[date.weekday - 1];
  final month = _monthNames[date.month - 1];

  return '$weekday ${date.day} $month';
}

// E.g. "Lun 3 ago".
String formatAvailableDayShortLabel(DateTime date)
{
  final weekday = _weekdayNames[date.weekday - 1].substring(0, 3);
  final month = _monthNames[date.month - 1].substring(0, 3);

  return '$weekday ${date.day} $month';
}
